renku/core/management/workflow/concrete_execution_graph.py

```diff
@@ -141,6 +141,9 @@ def workflow_graph(self):
                     workflow_graph.add_node(node)
                 continue
 
+            if not next(self.graph.predecessors(node), None):
```

Comment:
It might only make a difference on very large repositories with complex workflows, but 
```
intermediate_predecessor = next(self.graph.predecessors(node), None)
if not intermediate_predecessor:
    continue
[...]
source = next(self.graph.predecessors(intermediate_predecessor), None)
```
would only have to calculate the predecessor once.

Reply:
It always helps with the code's readability.